### PR TITLE
[SPARK-4819] Remove Guava's "Optional" from public API

### PR DESCRIPTION
**Pull request by srowen**
Replace Guava `Optional` with (an API clone of) Java 8 `java.util.Optional` (edit: and a clone of Guava `Optional`)

See also https://github.com/apache/spark/pull/10512

### CODE REVIEW

**Comment by vanzin** [comment]
This looks ok to me if we're ok with breaking source and binary compatibility in 2.0. I guess it's the less bad option to fix this without depending on Java 8...

**Comment by rxin** [comment]
If we are creating our own Optional, why don't we also support all the methods available on Guava, so we only need to break "import" compatibility?

**Comment by srowen** [comment]
My theory was that it makes it more compatible with Java 8 Optional. Later, moving to its implementation is a smaller change. We're taking the bigger change now. Or at least that was the motivation.

Another possibility is to add the Guava work-alike methods and deprecate them. Later they'd be removed I guess when Java 8 Optional gets used. I liked that less since it still just half of the breaking change farther down the road, when a major release seems like the time to do as much of the breaking as possible.

**Comment by rxin** [comment]
But we might never switch to Java 8's Optional, or maybe the entire API that's depending Optional will be deprecated in the future. Why should we increase the level of breakage now in anticipation of something that we don't even know we need in the future?

**Comment by srowen** [comment]
No matter what, there's a break now -- not even source-compatible. In for a penny, in for a pound, if it has a future benefit? but I agree it's unlikely to actually switch to the Java 8 `Optional` as it's yet another source-incompatible change. I think it's either use Java 8 in Spark 2.x, and then really use `java.util.Optional`, or else go with the hybrid solution where the new class tries to implement both the Guava and Java 8 API.

File: core/src/main/java/org/apache/spark/api/java/Optional.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+ */
+
+package org.apache.spark.api.java;
+
+import java.io.Serializable;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * API copied from {@code java.util.Optional} in Java 8 and
```
Comment (rxin) [inline core/src/main/java/org/apache/spark/api/java/Optional.java]: we should probably have a better user-facing documentation in addition to this ...

**Comment by rxin** [comment]
LGTM. Anything else you want to do? Otherwise we should remove WIP and merge it.

**Comment by srowen** [comment]
It's ready as far as I'm concerned.

**Comment by rxin** [comment]
Thanks - I'm going to merge this.

**Pull request closed without merging**